### ISSUE TITLE
How to Determine Sonos Model?

### ISSUE THREAD

**Issue by dundeemt** [SoCo/SoCo#320]
Looking to create a method to return the model type (PLAY:1, PLAY:3, PLAY:5, PLAYBAR, SUB, CONNECT, CONNECT:AMP, ??)

I have 3 PLAY:1 that all show a Hardware Version of 1.8.3.7-1  and a PLAYBAR with a hardware version of 1.9.1.10-1 -- if others could pull up their info and list as a comment:
  http://zone_ip:1400/support/review  and click on "Zone Player Info" and report back their Hardware version info along with the model, I believe we can create a method to return the model name based on it.  i.e.
  PLAYBAR - 1.19.10-1
  PLAY:1  - 1.8.3.7-1

I've google'd and couldn't find a resource but if I am missing something, please let me know (resource already available, etc)

Current Feedback as of 2015-07-24, 15:46 CST

|Model      |Hardware Versions
|---------     |-----------------------------
|PLAY:1    | **1.8.3.7-1**, 1.8.3.7-2
|PLAY:3    | 1.8.1.2-2
|PLAY:5    | **1.16.4.1-2**
|PLAYBAR| 1.9.1.10-1
|SUB          |
|CONNECT| **1.16.5.5-2**
|CONNECT:AMP|

If a hardware version has more than one person reporting the same version, it is in bold.  (One person reporting multiple matching versions is currently being counted as 1.  subject to feedback.)

**Comment by poirier** [comment]
Hardware version 1.8.3.7-1 - Play 1
1.16.5.5-1 - Connect

**Comment by DPH** [comment]
1.16.4.1-2 Play 5
1.16.5.5-2 Connect

Cheers David

**Comment by KennethNielsen** [comment]
```python
>>> for zone in soco.discover():
...     print(zone.player_name, zone.get_speaker_info()['hardware_version'])
... 
('Stue', '1.16.5.5-2')  # Connect
(u'K\xf8kken', '1.8.1.2-2')  # Play 3
(u'Sovev\xe6relse', '1.8.3.7-2')  # Play 1
(u'Badev\xe6relse', '1.8.3.7-2')  # Play 1
('Kontor', '1.16.4.1-2')  # Play 5
```

**Comment by KennethNielsen** [comment]
@dundeemt Maybe you cound create a table in the first post, so that we can easier see what we are missing.

**Comment by dundeemt** [comment]
@KennethNielsen an excellent suggestion.  done.

**Comment by the01** [comment]
It would be a lot easier and more reliable to simply parse <ip>:1400/xml/device_description.xml
There you get
```xml
<root>
  ...
  <device>
    ...
    <modelNumber>S3</modelNumber>
    <modelDescription>Sonos PLAY:3</modelDescription>
    <modelName>Sonos PLAY:3</modelName>
    <modelURL>http://www.sonos.com/products/zoneplayers/S3</modelURL>
    ...
  </device>
  ...
</root>
```
and take your pick on how to identify the model :)

**Comment by dundeemt** [comment]
Yes, it would.  I did not know about that.  I knew about the :1400/status
 - did you find this via /status or do you know of other query points under
/xml ?

Best,
Jeff

On Fri, Aug 21, 2015 at 3:54 PM, the01 <notifications@github.com> wrote:

> It would be a lot easier and more reliable to simply parse
> :1400/xml/device_description.xml
> There you get
>
> <root>
>   ...
>   <device>
>     ...
>     <modelNumber>S3</modelNumber>
>     <modelDescription>Sonos PLAY:3</modelDescription>
>     <modelName>Sonos PLAY:3</modelName>
>     <modelURL>http://www.sonos.com/products/zoneplayers/S3</modelURL>
>     ...
>   </device>
>   ...
> </root>
>
> and take your pick on how to identify the model :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/issues/320#issuecomment-133561557>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473

**Comment by the01** [comment]
No, I think that I found it on some [sonos forum](https://en.community.sonos.com/). There is great stuff on there. For example at support/review, you can check out the network matrix.
Loads of fun things to discover :smile:

**Comment by lawrenceakka** [comment]
We have some of this info at https://github.com/SoCo/SoCo/wiki/Information-Adresses , though that is rather out of date.

We really should replace the code for `get_speaker_info` so that it parses `device_description_xml` instead of `status/zp`, since this is part of the UPnP spec. We could then make the model name available in the returned speaker_info.

A nice job for someone who would like to create a pull request :wink:

**Comment by lawrenceakka** [comment]
Hmm ... the info from status/zp is also available as follows:
``` python
>>> print soco.deviceProperties.GetZoneInfo
```
It looks as though we did not convert the `get_speaker_info` code over to use the new UPnP methods.
The model name is not available there, though, so we may still need to parse `device_description_xml` to get at that.

**Comment by the01** [comment]
A preferred/recommended way of implementing this? I am not too familiar with the code base, but if you could give me some pointers, I'd give parsing `device_description.xml` a try?

**Comment by lawrenceakka** [comment]
>  I am not too familiar with the code base, 

Now's your chance!

Off the top of my head, what about replacing the code in `get_speaker_info` with something like this:

``` python
if self.speaker_info and refresh is False:
    return self.speaker_info
else:
    response = requests.get('http://' + self.ip_address +
                            ':1400/xml/device_description.xml')
    dom = XML.fromstring(response.content)

dev =  dom.find('.//{urn:schemas-upnp-org:device-1-0}device')
if dev is not None:
    self.speaker_info['zone_name'] = \
        dev.findtext('.//{urn:schemas-upnp-org:device-1-0}roomName')
.....

```
etc.  You may well find a better way to do it.  If you make a pull request, we'll be happy to give comments. Feel free to ask questions.

Don't forget unitises :wink:

Lawrence

**Comment by the01** [comment]
Ok, great, thanks! From a provisional look I thought there might be a way to do it with services or something, but your way should be pretty straight forward :smile:
Will have a look soon!

**Comment by the01** [comment]
Ok, I changed the function at the01@5a239d23c250de96d12f95acc83e657c8c50e0ea (just an initial commit). It would seem though that `zone_icon`and `mac_adress` are not present in `device_description.xml`. The MAC was not a big problem since it appears to be part of the serial number and I fixed it with 
```python
self.speaker_info['serial_number'].split(':')[0]
```

Should I add a second request to `status/zp` or do we simply deprecate this information?

**Comment by lawrenceakka** [comment]
Excellent.  I can see icon info under device/icon list/icon.  I would have thought that the url was the most important bit. I think we should use that, with a note saying that it has changed (hopefully, so that it is more useful)

**Comment by the01** [comment]
Yes, but it is a different icon. It used to be the zone icon and that one is for the player (e.g. s3)

**Comment by dundeemt** [comment]
I think this issue can be marked closed as of commit c0ada91defdc4051895d184421a2e0a3e91a884c
